### ISSUE TITLE
Energy parameter derivatives calculated incorrectly on CUDA under certain conditions

### ISSUE THREAD

**Issue by alexrd** [openmm/openmm#4247]
The conditions (too long for the issue title) are as follows:

1. multiple custom nonbonded forces are present
2. one or more of these uses an interaction group

I made a minimal example that demonstrates this using a system with two water molecules:

```python
n_part = 6

def addCNBtoSystem(system,idx):

    # get nbforce, change the method to NoCutoff, and copy the exceptions                                                                    
    nb_force = system.getForce(6)
    nb_force.setNonbondedMethod(nb_force.NoCutoff)

    num_except = nb_force.getNumExceptions()
    nb_exceptions = []
    for i in range(num_except):
        pars = nb_force.getExceptionParameters(i)
        nb_exceptions.append([pars[0],pars[1]])

    # build a new CustomNonbondedForce and attach to the system                                                                              
    energy_function = f'param{idx}'

    test_force = omm.CustomNonbondedForce(energy_function)
    test_force.addGlobalParameter(f'param{idx}',1.0)
    test_force.addEnergyParameterDerivative(f'param{idx}')

    test_force.setNonbondedMethod(test_force.NoCutoff)
    for i in range(n_part):
        test_force.addParticle()

    test_force.addInteractionGroup(set([0,1]),set([3]))

    for j in range(len(nb_exceptions)):
        test_force.addExclusion(nb_exceptions[j][0], nb_exceptions[j][1])

    test_force.setForceGroup(idx)

    system.addForce(test_force)
    return system
```

so `test_force` has an energy function that is simply the `param1` or `param2` value, which we set to 1.  That way both the energy and the EnergyParameterDerivative are just counting interactions.  Since we added an interaction group of (0,1) and (3), each force should calculate 2 interactions.  The bug manifests itself when there are at least two of these forces on the same system.

I won't post all the code here ([see this zip file](https://github.com/openmm/openmm/files/12731988/energy_param_deriv_example.zip)), but when these are run with the Reference platform we get the expected output:

```bash
bash-5.1$ python test_deriv_two_waters.py Reference
Using Reference platform..
Deriv. of energy parameter with one force:
Param derivative 1: 2.0
Energy1: 2.000409091217244 kJ/mol
Deriv. of energy parameters with two forces:
Param derivative 1: 2.0
Param derivative 2: 2.0
Energy1: 2.000409091217244 kJ/mol
Energy2: 2.0 kJ/mol
```

with CUDA, the energies are correct, but the EnergyParameterDerivatives are messed up:

```bash
bash-5.1$ python test_deriv_two_waters.py CUDA
Using CUDA platform..
Deriv. of energy parameter with one force:
Param derivative 1: 2.0
Energy1: 2.0004093647003174 kJ/mol
Deriv. of energy parameters with two forces:
Param derivative 1: 1.0
Param derivative 2: 3.0
Energy1: 2.0004093647003174 kJ/mol
Energy2: 2.0 kJ/mol
```

This does not occur when the interaction groups are absent. Seems like the energy parameter derivative contributions are being allocated to the wrong parameters?

**Comment by peastman** [comment]
Thanks for the excellent test case!  I think I see what the problem is.  I'll try to get a fix in shortly.

**Comment by peastman** [comment]
The fix is in #4249.

**Comment by alexrd** [comment]
Thanks @peastman !  The patched version works for us as well.
